indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java

```diff
@@ -111,32 +125,152 @@ public static void setupClasspath(
     }
 
     for (String jarFilePath : jarFiles) {
-      File jarFile = new File(jarFilePath);
+
+      final File jarFile = new File(jarFilePath);
       if (jarFile.getName().endsWith(".jar")) {
-        final Path hdfsPath = new Path(distributedClassPath, jarFile.getName());
-
-        if (!existing.contains(hdfsPath)) {
-          if (jarFile.getName().matches(".*SNAPSHOT(-selfcontained)?\\.jar$") || !fs.exists(hdfsPath)) {
-            log.info("Uploading jar to path[%s]", hdfsPath);
-            ByteStreams.copy(
-                Files.newInputStreamSupplier(jarFile),
-                new OutputSupplier<OutputStream>()
+        try {
+          RetryUtils.retry(
+              new Callable<Boolean>()
+              {
+                @Override
+                public Boolean call() throws Exception
                 {
-                  @Override
-                  public OutputStream getOutput() throws IOException
-                  {
-                    return fs.create(hdfsPath);
+                  if (isSnapshot(jarFile)) {
+                    addSnapshotJarToClassPath(jarFile, intermediateClassPath, fs, job);
+                  } else {
+                    addJarToClassPath(jarFile, distributedClassPath, intermediateClassPath, fs, job);
                   }
+                  return true;
                 }
-            );
-          }
+              },
+              shouldRetryPredicate(),
+              DEFAULT_RETRY_COUNT
+          );
+        }
+        catch (Exception e) {
+          throw Throwables.propagate(e);
+        }
+      }
+    }
+  }
 
-          existing.add(hdfsPath);
+  public static final Predicate<Throwable> shouldRetryPredicate()
+  {
+    return new Predicate<Throwable>()
+    {
+      @Override
+      public boolean apply(Throwable input)
+      {
+        if (input == null) {
+          return false;
+        }
+        if (input instanceof IOException) {
+          return true;
         }
+        return apply(input.getCause());
```

Comment:
addJarToClassPath has 
```
throw new ISE("File does not exist even after moving from[%s] to [%s]", intermediateHdfsPath, hdfsPath);
```

that means it won't be retried. Is that  intentional?

why not retry here irrespective of what the exception type is, in the worst case there would be 3 retries instead of having to do a recursion?